applications/OptimizationApplication/python_scripts/utilities/buffered_dict.py

```diff
@@ -0,0 +1,406 @@
+from __future__ import annotations
+from typing import Any
+
+class BufferedDict:
+    """Buffered dict container with buffered data ability
+
+    Instances of this can hold (str, Any) data pairs in hierachychal data
+    structure where each sub item can have their own buffer sizes. Hence,
+    Different sub structures can have their own buffers and can be advanced
+    seperately or all together.
+
+    Overwriting of the existing data is not allowed because, this is used
+    as a central place to store data from various objects, hence allowing
+    overwriting will make difficult to track the changes.
+
+    Buffer size setting can only be done at the construction time
+    of the instance to not to allow re sizing because then it is not
+    clear how to handle the connected steps.
+
+    ClearStep can be used to clear all the data in the specified step_index.
+
+    The buffered data is stored in a cyclic buffer.
+    """
+    def __init__(self, buffer_size: int = 1):
+        """Creates an instance of BufferedDict with specified buffer size
+
+        This creates an instance of buffered data with given buffer size.
+
+        Args:
+            buffer_size (int, optional): Cyclic buffer size. Defaults to 1.
+        """
+        self.__parent = None
```

Comment:
@sunethwarna I understand the benefits of and needs for this class however, I am sure it can be simplified. It is too complicated and not friendly to be used. I explain it in the following. 
Basically, this class is containing historical data in step fashion, so there is a list of dicts, called __buffered_data which has a fixed size of buffer_size. Since the size is fixed, the __buffer_index and functions like  AdvanceStep become completely redundant. If the buffered data is ordered according the step index, i.e.,  _buffered_data = [step_0_dict,step_1_dict,step_2_dict,..], one can access everything without calling  AdvanceStep and using __buffer_index. In the current implementation since step_0 which is the latest state of the data, refers to the last item of the  _buffered_data one needs to use __buffer_index. However if we order the data like  _buffered_data = [step_0_dict,step_1_dict,step_2_dict,..], we are fully consistent and it is super handy to use. So I am sure we need to simplify the class as explained.